<HTML>
<HEAD>
<TITLE>
Quality Control Reports
</TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF">

<HR NOSHADE>
<H1>MGI Quality Control Reports</H1>
<HR>
QC reports are generated nightly from the MGD Production Database (unless otherwise noted).
They are intended for internal use only (i.e. not available for public consumption, although they may 
be used at times to test reports for <I>future</I> public consumption.)
<HR>

<B>Table of Contents</B>
<UL>
<LI><A HREF="#references">References</A>
<LI><A HREF="#markers">Markers</A>
<LI><A HREF="#alleles">Alleles</A>
<LI><A HREF="#phenoannot">Phenotype Vocabulary Annotation</A>
<LI><A HREF="#mlc">MLC</A>
<LI><A HREF="#probes">Probes</A>
<LI><A HREF="#mapping">Mapping</A>
<LI><A HREF="#orthology">Orthology</A>
<LI><A HREF="#gxd">GXD</A>
<LI><A HREF="#go">GO</A>
<LI><A HREF="#nomenclature">Nomenclature</A>
<LI><A HREF="#strains">Strains</A>
<LI><A HREF="#mgs">MGS</A>
<LI><A HREF="#sequence">Sequence</A>
<LI><A HREF="#dataloads">Data Loads</A>
<LI><A HREF="#lists">Lists/Other</A>
<LI><A HREF="ftp://ftp.informatics.jax.org/pub/reports/index.html">Public Reports</A>
</UL>
<HR>

<A NAME="references"><B>References</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/BIB_NeedNLM1.sql.rpt">References Which Need NLM Updates</A>
<LI><A HREF="/data/reports/qcreports_db/output/BIB_InPress.sql.rpt">References In Press/Submissions</A>
<LI><A HREF="/data/reports/qcreports_db/output/BIB_NoAbstract.sql.rpt">References w/ Medline UI and No Abstract</A>
</OL>

<A NAME="markers"><B>Markers</B> </A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_OffsetsMLDP.sql.rpt">Markers with No Offsets Modifed in Mapping w/in the last 3 days</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Offsets.sql.rpt">Markers w/ MGD Offset but w/out CC Offset or Mapping Information</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_OffsetsUN.sql.rpt">Markers with Offsets but no Chromosome Assignment</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Matrix.sql.rpt">UN/Syntenic Markers In MGD found in Mapping Panel Crosses</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_LowerSeq.rpt">Lowercase Gene Symbols which have GenBank IDS</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_AssigningRef.rpt">MGI Markers with Assigning Reference</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_MGC.sql.rpt">MGC IDs with Nomen Records Not Deleted or Broadcast</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoSequence.rpt">Genes with No Sequence Data, Excluding Mutant Phenotypes</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_QTL.rpt">QTL Markers added to MGI since March 2002 with no PhenoSlim annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_SharedSynonyms.rpt">Markers with Shared Synonyms</A>
</OL>

<A NAME="alleles"><B>Alleles</B> </A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Alleles.sql.rpt">Allele Anomalies</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_ESCellStrain.sql.rpt">ES Cell Line Strains Modified Last Month - Sorted by Modification Date</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_WithAllele_MLC.sql.rpt">Allele Withdrawals in Last 2 Months where either Old or New Gene has MLC record</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_Progress.sql.current.rpt">Allele Progress Report</A>
<LI><A HREF="/data/reports/qcreports_db/archive/alleles">Allele Progress Archive</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_NoMutantSentence.rpt">Genes w/Phenotypic Alleles and No Mutant Sentence</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_NoAnnotation.sql.rpt">Alleles w/ 'Associated Phenotype Controlled Terms' in Notes and no PhenoSlim or MP annotation</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_MarkerClip.sql.current.rpt">Genes with Marker Clips and Alleles that have been added in the last 5 days</A>
</OL>

<A NAME="phenoannot"><B>Phenotype Vocabulary Annotation</B> </A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_ImmuneAnnot.current.html">Alleles with Immune System Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_MPObsolete.sql.rpt">Genotype Annotations to Obsolete MP Terms</A>
</OL>

<A NAME="mlc"><B>MLC</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_MLC_Withdrawn.sql.rpt">Withdrawn Markers w/ MLC Text</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLC_Alleles.sql.rpt">MLC Symbols which contain Alleles</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLC_References.sql.rpt">References which only appear in MLC</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLC_CheckImageIndex.rpt">MLC Image Index File Entries Which Contain Withdrawn Symbols</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLCNomenclature-current.html">Updates to MLC Mouse Nomenclature</A>
</OL>

<A NAME="probes"><B>Probes</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Reference.sql.rpt">Probes/Primers w/out References</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Marker.sql.rpt">Probes Used More Than Once for Marker/Duplicate Markers/No Markers (excluding IMAGE, RPCI, NIA, RIKEN clones)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_NoGenBank.sql.rpt">No GenBank IDs - Sequence in Paper</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_NoMarkerSeq.sql.rpt">Molecular Segments with Encoding Markers and Sequence IDs and no corresponding Marker-Sequence ID association</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_MarkerSeq.sql.rpt">IMAGE Clone/Marker Sequence Anomalies</A>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_ProblemSequence.rpt">Molecular Segment with the Problem Sequence Note associated with more than one GenBank Seqid or associated with no GenBank Seqid</A>
<LI><A HREF="ftp://ftp.informatics.jax.org/pub/reports/MGI_ProblemSequence.rpt">Public Problem Sequence Report</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_RIKENMarkers.rpt">RIKEN Clones and MGI Markers</A>
</OL>

<A NAME="mapping"><B>Mapping</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MLD_Chromosome.sql.rpt">Invalid Chromosomes</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLD_2Point.sql.rpt">Invalid Recombinants vs. Parentals/Recombinants vs. Total</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLD_Cross.sql.rpt">Invalid Cross Data</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLD_Marker.sql.rpt">Missing Experimental Markers</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLD_DupMarker.sql.rpt">Duplicate Markers Appearing in Experiments</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLD_ExptMismatch.sql.rpt">Experiments which are of one Experiment Type but which have corresponding data records in non-related Experiment tables</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLD_NomenText.sql.rpt">Symbols which have undergone withdrawals and appear in Text Experiments</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_MappedNoRef.sql.rpt">Mapped Markers w/out a Mapping Reference</A>
</OL>

<A NAME="orthology"><B>Orthology</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_References2.sql.rpt">References selected for Orthology only w/out Orthology Data</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_References5.sql.rpt">References selected for Orthology w/ no Orthology Data with at least one reference-to-gene Association</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_SpeciesCount.sql.rpt">Counts of Mouse Orthologies by Organism</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_Duplicates.sql.rpt">Duplicate Orthologies (by Class, Reference, Species)</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_ConservedUN.sql.rpt">Orthologies w/ Conserved Map Location and Chromosome = UN</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_NoMouseSeq.sql.rpt">Mouse Markers in Orthology w/ Nucleotide or Amino Acid Sequence Similarity Assay and Marker has no Seq ID</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_RIKEN.sql.rpt">Human/Mouse Orthologies involving RIKEN Symbols</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_MouseSeq1.rpt">Mouse Genes with Sequence ID but no Human Orthology</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_MouseSeq2.sql.rpt">Mouse Genes with no Sequence ID but with Human Orthology</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_NoMouse.sql.rpt">Orthology Records w/out Mouse Genes</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_HomoloGene1.sql.rpt">HomoloGene Mouse/Human Orthologous Pairs where we have neither the mouse nor the human member of the pair. (HUMAN AND MOUSE SYMBOLS ARE THE SAME)</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_HomoloGene2A.sql.rpt">HomoloGene Mouse/Human Orthologous Pairs where we have neither the mouse nor the human member of the pair. (HUMAN AND MOUSE SYMBOLS ARE NOT THE SAME, HUMAN SYMBOLS ARE OFFICIAL)</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_HomoloGene2B.sql.rpt">HomoloGene Mouse/Human Orthologous Pairs where we have neither the mouse nor the human member of the pair. (HUMAN AND MOUSE SYMBOLS ARE NOT THE SAME, HUMAN SYMBOLS ARE INTERIM)</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_HomoloGene3.sql.rpt">HomoloGene Mouse/Human Orthologous Pairs where our Orthology contains Human but no Mouse member.</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_HomoloGene4.sql.rpt">HomoloGene Mouse/Human Orthologous Pairs where our Orthology contains Mouse but no Human member.</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_Singletons.sql.rpt">Mouse/Human Singletons</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_RatNoSeq.sql.rpt">Rat Genes with no Sequence ID but with Orthology</A>
</OL>

<A NAME="gxd"><B>GXD</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ProbeMarker.sql.rpt">GXD Assay Probe/Marker Pairs No Longer Found in Master Probe Table</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_AntibodyMarker.sql.rpt">GXD Assay Antibody/Marker Pairs No Longer Found in Master Antibody Table</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ProbeReference.sql.rpt">GXD Assay Probes with no corresponding entry in the Probe Reference Table</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_NotInCache.sql.rpt">GXD Assays missing from GXD Expression Cache Table</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_Stats.sql.rpt">GXD Statistics</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SpecEmbryonic.sql.rpt">Adult Specimens with Embryonic Structures</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SpecNoAge.sql.rpt">Specimens with NULL Age Value (when there should be one)</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SpecMissingAllele.sql.rpt">Specimens with Missing Allele 2 Genotypes</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SpecTheiler.sql.rpt">Specimens annotated to structures of > 1 Theiler Stage</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_MechDev.rpt">Mech Dev papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_GeneExprPatterns.rpt">Gene Expr Patterns papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_DevBiol.rpt">Dev Biol papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_Development.rpt">Development papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_DevDyn.rpt">Dev Dyn papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SpecTheilerAge.rpt">GXD Specimens with incompatible Theiler stages and ages</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_KnockInGene.sql.rpt">GXD Knock Ins where the assayed Gene is not the Gene which is mutated in the Genotype</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_GenotypeRef.sql.rpt">References where GXD and Pheno differ in Genotype</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_WildType.sql.rpt">Genotypes where Allele 1 is "wild type"</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_GenotypeNA.sql.rpt">GXD Annotations to "Not Applicable" Genotype</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_Images.rpt">Papers Requiring Images</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_LackingImages.rpt">Papers Requiring Permissions</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_DupGenotype.rpt">Duplicate Genotypes</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ImageCheck.sql.rpt">GXD Image Check</A>
</OL>

<A NAME="go"><B>GO</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_RatNoGO.sql.rpt">Mouse Genes that have Rat Homologs but no GO associations</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoGO_A.rpt">Genes (neither 'RIKEN' or 'expressed' or 'EST') with no GO Associations and reference count excludes data loads and GenBank submissions</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoGO_B.rpt">Genes (neither 'RIKEN' or 'expressed' or 'EST') with no GO Associations</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoGO_C.rpt">Genes (neither 'RIKEN' or 'expressed' or 'EST') with no GO Associations and reference count includes only data loads and GenBank submissions</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoGO_D.html">Genes (neither 'RIKEN' or 'expressed' or 'EST') with no GO Associations with references that are selected for GO but have not been used</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_AllNoGO.html">Genes with references that are selected for GO but have not been used</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOIEA_A.rpt">Genes (neither 'RIKEN' or 'expressed' or 'EST') with GO Associations of evidence IEA only and reference count excludes data loads and GenBank submissions</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOIEA_B.rpt">Genes (neither 'RIKEN' or 'expressed' or 'EST') with GO Associations of evidence IEA only</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOIEA_C.rpt">Genes (neither 'RIKEN' or 'expressed' or 'EST') with GO Associations of evidence IEA only and reference count includes only data loads and GenBank submissions</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOIEA_D.html">Genes (neither 'RIKEN' or 'expressed' or 'EST') with GO Associations of evidence IEA only and with references that are selected for GO but have not been used</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOGold_A.rpt">All Genes with Annotations to IDA, IGI, IMP, IPI, TAS</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOGold_B.rpt">All Genes with Annotations to IDA, IGI, IMP, IPI</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOGold_C.rpt">All Genes with multiple Annotations to IDA, IGI, IMP, IPI, TAS</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOGold_D.rpt">All Genes with multiple Annotations to IDA, IGI, IMP, IPI</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOAllele.rpt">Genes with Annotation to IMP, Inferred From = NULL and Reference Associations to Alleles</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_EvidenceCodeStats.sql.rpt">Total # of Annotations by GO Evidence Code</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_InvalidInferred.rpt">Invalid "Inferred From" Values in GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_stats.rpt">GO Statistics</A>
<LI><A HREF="/data/reports/qcreports_db/archive/go">Archive of GO Statistics</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOUnknown.sql.rpt">All genes with 'unknown' annotations with new indexed literature</A>
</OL>

<A NAME="nomenclature"><B>Nomenclature</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MLP_Nomenclature.sql.rpt">Strains Affected by recent Nomenclature changes</A>
<LI><A HREF="/data/reports/qcreports_db/output/NOMEN_Broadcast.sql.rpt">Nomenclature Symbols Broadcast Within Last 3 Days</A>
<LI><A HREF="/data/reports/qcreports_db/output/NOMEN_Pending.sql.rpt">Nomenclature In Progress Symbols</A>
<LI><A HREF="/data/reports/qcreports_db/output/NOMEN_Reserved.rpt">Nomenclature Reserved</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_SymbolDiffs1.sql.rpt">MGD Mouse Symbols which differ from Orthologous Human Symbols #1; excludes RIKEN</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_SymbolDiffs2.sql.rpt">MGD Mouse Symbols which differ from Orthologous Human Symbols #2; excludes RIKEN</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_SymbolDiffs3.sql.rpt">MGD Mouse Symbols which differ from Orthologous Human Symbols #3; RIKEN only</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_SymbolDiffs4.sql.rpt">MGD Mouse Symbols which differ from Orthologous Human Symbols #4; RIKEN only</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Withdrawn.sql.rpt">Withdrawn Markers</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_InterimByRef.sql.rpt">Interim Markers (sorted by Reference)</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_InterimByDate.sql.rpt">Interim Markers (sorted by Date)</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NomenRat.sql.rpt">Mouse Symbols with nomenclature changes which contain Rat homologs with unofficial nomenclature</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_ESTSynonyms.sql.rpt">EST IDs and Gm Markers as Synonyms</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NomenNotInformative.sql.rpt">MGI Markers with Uninformative Nomenclature #1</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NomenNotInformative2.sql.rpt">MGI Markers with Uninformative Nomenclature #2</A>
</OL>

<A NAME="strains"><B>Strains</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainStd.sql.rpt">Strains (Standard only)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainNonStd.sql.rpt">Strains (Non Standard excluding F1 and F2)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainF1F2.sql.rpt">Strains Containing F1 or F2</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainDup.sql.rpt">Duplicate Strains</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainByDate.sql.rpt">Strains (Non Standard - Sorted by Creation Date)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_NameChanges.rpt">Strain Names Which Have Been Changed</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain.rpt">Tab-delimited File of All Strains (MGI Acc ID, Strain, Markers, Synonyms, External Acc IDs)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain2.rpt">Tab-delimited File of All Strains (Strain/Standard)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainTrans.sql.rpt">Strains - Transgenic or Targeted Mutation</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainNonStdNoData.sql.rpt">Non-Standard Strains with no JR# and no data attached</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_MultNoAlleles.sql.rpt">Mutant Strains w/ Multiple Markers and No Alleles</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_SingleNoAlleles1.sql.rpt">Mutant Strains w/ Single Markers, No Alleles and Allele Symbol is in MGI</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_SingleNoAlleles2.sql.rpt">Mutant Strains w/ Single Markers, No Alleles and Allele Symbol is not in MGI</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_NoAlleles1.sql.rpt">Strains w/ Markers, No Alleles and Allele Symbol is in MGI</A>
</OL>

<A NAME="mgs"><B>MGS</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_GenBank.rpt">All Sequences with Marker/Molecular Segment (bobs)</A>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_Marker.rpt">Sequences Annotated to Mouse Markers (excludes TIGR, DoTS, NIA); tab-delimited</A>
<LI><A HREF="/data/reports/qcreports_db/output/ACC_GenBank.sql.rpt">All GenBank IDs in MGI Associated with a Molecular Segment or Mouse Gene</A>
</OL>

<A NAME="sequence"><B>Sequences</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_Deleted.sql.rpt">Deleted Sequences with MGI Associations</A>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_SeqRefMrkNoSeqMrk.sql.rpt">Sequence References Associated with Markers but no Sequence/Marker</A>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_Split.sql.rpt">Split Sequences</A>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_MarkerAnnotation.sql.rpt">Markers Annotated to a Secondary Sequence Accession ID</A>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_CloneSource.sql.rpt">Sequences and their Molecular Segments Associated with Different Sources</A>
</OL>

<A NAME="dataloads"><B>Data Loads</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/live/dataload/entrezgeneload/index.html">EntrezGene Load: Data and QC Reports</A>
<LI><A HREF="/usrlocalmgi/live/dataload/swissload">SwissProt Load Data and Logs</A>
<LI>UniGene</LI>
	<OL>
	<LI><A HREF="/data/reports/qcreports_db/output/multipleClusters.rpt">MGI Genes associated with multiple UniGene Clusters</A>
	<LI><A HREF="/data/reports/qcreports_db/output/multipleGenes.rpt">UniGene Clusters associated with multiple MGI Genes</A>
	<LI><A HREF="/data/reports/qcreports_db/output/genesMissingUG.rpt">MGI mouse Genes with Seq. ID but not in a UniGene Cluster</A>
	</OL>
<LI><A HREF="/usrlocalmgi/live/dataload/crossload/crossloadindex.html">CrossLoad: Data and QC Reports</A>
<LI><A HREF="/usrlocalmgi/live/dataload/vocload/GOindex.html">GO Vocabulary/DAG Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/vocload/MPindex.html">Mammalian Phenotype Vocabulary/DAG Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/vocload/MAindex.html">Adult Mouse Anatomy Vocabulary/DAG Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/expseqload/index.html">RH-Mapped Expressed Sequence Load (Lucy Load)</A>
<LI><A HREF="/usrlocalmgi/live/rpci/rpciload/index.html">RPCI Clone Load</A>
<LI><A HREF="/usrlocalmgi/live/genbank/gbseqload/index.html">GenBank Sequence Load</A>
<LI><A HREF="/usrlocalmgi/live/genbank/refseqload/index.html">RefSeq Sequence Load</A>
<LI><A HREF="/usrlocalmgi/live/swissprot/spseqload/index.html">SwissProt/TrEMBL Sequence Load</A>
<LI><A HREF="/usrlocalmgi/live/tigr/tigrseqload/index.html">TIGR Sequence Load</A>
<LI><A HREF="/usrlocalmgi/live/dots/dotsseqload/index.html">DoTS Sequence Load</A>
<LI><A HREF="/usrlocalmgi/live/nia/niaseqload/index.html">NIA Sequence Load</A>
</OL>


<A NAME="lists"><B>Lists/Other</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Tissue.sql.rpt">Tissues (All)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_CellLine.sql.rpt">Cell Lines</A>
<LI><A HREF="/data/reports/qcreports_db/output/ACC_ActualDB.sql.rpt">Logical/Actual DBs</A>
<LI><A HREF="/data/reports/qcreports_db/output/ACC_Accession.sql.rpt">Invalid EntrezGene/RefSeq IDs</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_OrphanGenotype.sql.rpt">Orphan Genotypes</A>
</OL>

<HR>

</BODY>
</HTML>
