<HTML>
<HEAD>
<TITLE>
Quality Control Reports
</TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF">

<HR NOSHADE>
<H1>MGI Quality Control Reports</H1>
<HR>
QC reports are generated nightly from the MGD Production Database (unless otherwise noted).
They are intended for internal use only (i.e. not available for public consumption, although they may 
be used at times to test reports for <I>future</I> public consumption.)
<HR>

<B>Table of Contents</B>
<UL>
<LI><A HREF="#references">References</A>
<LI><A HREF="#markers">Markers</A>
<LI><A HREF="#mlc">MLC</A>
<LI><A HREF="#probes">Probes</A>
<LI><A HREF="#mapping">Mapping</A>
<LI><A HREF="#homology">Homology</A>
<LI><A HREF="#gxd">GXD</A>
<LI><A HREF="#nomenclature">Nomenclature</A>
<LI><A HREF="#dataloads">Data Loads</A>
<LI><A HREF="#lists">Lists</A>
<LI><A HREF="#other">Other</A>
<LI><A HREF="ftp://ftp.informatics.jax.org/pub/informatics/reports/index.html">Public Reports</A>
<LI><A HREF="ftp://ftp.informatics.jax.org/pub/informatics/reports/tab-delimited.html">Tab-Delimited Reports</A>
</UL>
<HR>

<A NAME="references"><B>References</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/BIB_NeedNLM.sql.rpt">References Which Need NLM Updates</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/BIB_InPress.sql.rpt">References In Press/Submissions</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/BIB_NoAbstract.sql.rpt">References w/ Medline UI and No Abstract</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/BIB_Submission.sql.rpt">Potential Submission Matches</A>
</OL>

<A NAME="markers"><B>Markers</B> </A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_OffsetsMLDP.sql.rpt">Markers with No Offsets Modifed in Mapping w/in the last 3 days</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Offsets.sql.rpt">Markers w/ MGD Offset but w/out CC Offset or Mapping Information</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_OffsetsUN.sql.rpt">Markers with Offsets but no Chromosome Assignment</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Alias.sql.rpt">Aliases</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_NoReferences.sql.rpt">Markers w/out ANY References</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_References.sql.rpt">Markers w/out History References or History</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Withdrawn.sql.rpt">Splits and Withdrawn Markers w/ Additional Info</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Matrix.sql.rpt">UN/Syntenic Markers In MGD found in Mapping Panel Crosses</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Alleles.sql.rpt">Markers Which Contain Alleles</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_AllelesByDate.sql.rpt">Markers Which Contain Alleles (By Modification Date)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_AllelesAll.sql.rpt">Markers and Alleles</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_AccDup.sql.rpt">Duplicate Marker Accession Numbers (Sequence, SWISS-PROT)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_UniGeneClusterDiscrep.sql.rpt">Clusters with Markers on different chromosomes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_LowerSeq.rpt">Lowercase Gene Symbols which have GenBank IDS</a>
</OL>

<A NAME="mlc"><B>MLC</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_MLC_Withdrawn.sql.rpt">Withdrawn Markers w/ MLC Text</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLC_Alleles.sql.rpt">MLC Symbols which contain Alleles</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLC_References.sql.rpt">References which only appear in MLC</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLC_CheckImageIndex.rpt">MLC Image Index File Entries Which Contain Withdrawn Symbols</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLC_TransferStats.rpt">MLC Transfer Statistics Report</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLCNomenclature-current.html">Updates to MLC Mouse Nomenclature</A>
</OL>

<A NAME="probes"><B>Probes</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_Reference.sql.rpt">Probes/Primers w/out References (excluding I.M.A.G.E. clones)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_Marker.sql.rpt">Probes Used More Than Once for Marker/Duplicate Markers/No Markers (excluding I.M.A.G.E. clones)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_NoGenBank.sql.rpt">No GenBank IDs - Sequence in Paper</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_NoMarkerSeq.sql.rpt">Molecular Segments with Encoding Markers and Sequence IDs and no corresponding Marker-Sequence ID association</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_MarkerSeq.sql.rpt">I.M.A.G.E. Clone/Marker Sequence Anomalies</A>
</OL>

<A NAME="mapping"><B>Mapping</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_Chromosome.sql.rpt">Invalid Chromosomes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_2Point.sql.rpt">Invalid Recombinants vs. Parentals/Recombinants vs. Total</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_Cross.sql.rpt">Invalid Cross Data</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_Marker.sql.rpt">Missing Experimental Markers</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_DupMarker.sql.rpt">Duplicate Markers Appearing in Experiments</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_ExptMismatch.sql.rpt">Experiments which are of one Experiment Type but which have corresponding data records in non-related Experiment tables</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_NomenText.sql.rpt">Symbols which have undergone withdrawals and appear in Text Experiments</A>
</OL>

<A NAME="homology"><B>Homology</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_References2.sql.rpt">References selected for Homology only w/out Homology Data</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_References3.sql.rpt">References selected for Homology and MLC only w/out Homology Data</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_References4.sql.rpt">References selected for Homology and Probes w/out Homology Data</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_References5.sql.rpt">References selected for Homology w/ 1 Gene</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_Unreviewed.sql.rpt">Homologies for Oxford Grid Species w/ Unreviewed Assay</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_SpeciesCount.rpt">Counts of Mouse Homologies by Species</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_Duplicates.sql.rpt">Duplicate Homologies (by Class, Reference, Species)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_ConservedUN.sql.rpt">Homologies w/ Conserved Map Location and Chromosome = UN</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_Potentials.html">MGD Entries with Potential Human Homologies</A>
</OL>

<A NAME="gxd"><B>GXD</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_ProbeMarker.sql.rpt">GXD Assay Probe/Marker Pairs No Longer Found in Master Probe Table</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_ProbeReference.sql.rpt">GXD Assay Probes with no corresponding entry in the Probe Reference Table</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_AltTranscripts.sql.rpt">Alterntative Transcripts that have "encodes" relationship for more than one gene</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_NotInCache.sql.rpt">GXD Assays missing from GXD Expression Cache Table</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_Stats.sql.rpt">GXD Statistics</A>
</OL>

<A NAME="nomenclature"><B>Nomenclature</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLP_Nomenclature.sql.rpt">Strains Affected by recent Nomenclature changes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/NOMEN_Broadcast.sql.rpt">Nomenclature Symbols Broadcast Within Last 3 Days</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/NOMEN_Pending.sql.rpt">Nomenclature Pending Symbols</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/NOMEN_Duplicate.sql.rpt">Nomenclature Duplicate Symbols</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/NOMEN_Reserved.rpt">Nomenclature Reserved</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Duplicates.sql.rpt">MGD Duplicate Markers</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_SymbolDiffs.sql.rpt">MGD Mouse Symbols which differ from Orthologous Human Symbols</A>
</OL>

<A NAME="dataloads"><B>Sequence DB Loads</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/dataload/locuslinkload/LLindex.html">LocusLink Load: Data and QC Reports</A>
<LI><A HREF="/usrlocalmgi/dataload/gdbload/GDBindex.html">GDB Load: Data and QC Reports</A>
<LI><A HREF="/usrlocalmgi/dataload/ratmapload/RATMAPindex.html">RATMAP Load: Data and QC Reports</A>
<LI><A HREF="/usrlocalmgi/dataload/swissload">SwissProt Load Data and Logs</A>
<LI>UniGene</LI>
	<OL>
	<LI><A HREF="/usrlocalmgi/qcreports_db/output/multipleClusters.rpt">MGI Genes associated with multiple UniGene Clusters</A>
	<LI><A HREF="/usrlocalmgi/qcreports_db/output/multipleGenes.rpt">UniGene Clusters associated with multiple MGI Genes</A>
	<LI><A HREF="/usrlocalmgi/qcreports_db/output/genesMissingUG.rpt">MGI mouse Genes with Seq. ID but not in a UniGene Cluster</A>
	</OL>
</OL>


<A NAME="lists"><B>Lists</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Species.sql.rpt">Marker Species</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Classes.sql.rpt">Marker Classes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_Assays.sql.rpt">Mapping Assays</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_Crosses.sql.rpt">Mapping Panels Displayed in Lookup List</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_Strain.sql.rpt">Strains (All)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_StrainNonStd.sql.rpt">Strains (Non Standard)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_StrainNonStdUpdate.sql.rpt">Strains (Non Standard) Sorted by Creation Date</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_Tissue.sql.rpt">Tissues (All)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_CellLine.sql.rpt">Cell Lines</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_TissueCell.sql.rpt">Unique Tissue/Cell Lines Pairs</A>
</OL>

<A NAME="other"><B>Other</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/ACC_Accession.sql.rpt">Records missing MGI Accession Numbers</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_StrainDup.sql.rpt">Duplicate Strains</A>
</OL>

<HR>

</BODY>
</HTML>
