<HTML>
<HEAD>
<TITLE>
Quality Control Reports
</TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF">

<HR NOSHADE>
<H1>MGI Quality Control Reports</H1>
<HR>
QC reports are generated nightly from the MGD Production Database (unless otherwise noted).
They are intended for internal use only (i.e. not available for public consumption, although they may
be used at times to test reports for <I>future</I> public consumption.)
<HR>

<B>Table of Contents</B>
<UL>
<LI><A HREF="#references">References</A>
<LI><A HREF="#alleles">Alleles</A>
<LI><A HREF="#phenoannot">Phenotype Vocabulary Annotation</A>
<LI><A HREF="#gxd">GXD</A>
<LI><A HREF="#recomb">Recombinase and Transgenic Reporter Data</A>
<LI><A HREF="#go">GO</A>
<LI><A HREF="#mtb">MTB</A>
<LI><A HREF="#nomenclature">Nomenclature</A>
<LI><A HREF="#strains">Strains</A>
<LI><A HREF="#strainsjax">Strains - JAX</A>
<LI><A HREF="#sequencegroup">Sequence Group</A>
<LI><A HREF="#sequence">Sequence</A>
<LI><A HREF="#homology">Homology</A>
<LI><A HREF="#dataloads">Data Loads</A>
<LI><A HREF="ftp://ftp.informatics.jax.org/pub/reports/index.html">Public Reports</A>
</UL>
<HR>

<A NAME="references"><B>References</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/BIB_NoAbstract.sql.rpt">References w/ PubMed ID and No Abstract</A> <B>(jeo)</B>
</OL>

<A NAME="alleles"><B>Alleles (csmith, smb, dlb, honda, mnk)</B> </A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Alleles.sql.rpt">Allele Anomalies</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_Progress.current.rpt">Allele Progress Report</A>
<LI><A HREF="/data/reports/qcreports_db/archive/alleles">Allele Progress Archive</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_NoMutantSentence.rpt">Genes w/Phenotypic Alleles and No Mutant Sentence</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_DupGenotype.rpt">Duplicate Genotypes</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_NoMolMutation.rpt">Alleles that have Molecular Notes but no Molecular Mutation</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_MolNotesNoMP.rpt">Alleles that have Molecular Notes but no MP Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_WildType.sql.rpt">Genotypes where Allele 1 is "wild type"</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_CellLineProvider.sql.rpt">Mutant ES Cell Line Provider Check</A>
<LI><A HREF="/data/reports/qcreports_db/output/PHENO_ImageCheck.sql.rpt">Image Check</A>
<LI><A HREF="/data/reports/qcreports_db/output/PHENO_Images.rpt">Phenotype Annotations Requiring Images</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_SequenceReference.sql.rpt">Reference associations created by the gene trap load</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_GermlineTransmission.sql.rpt">Germline Transmission Checks</A>
<LI><A HREF="/data/reports/qcreports_db/output/allstrain.rpt">Allele Strain of Origin vs. Parent Cell Line Strain</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_AllelePairCheck.sql.rpt">Genotypes where Allele Pairs are associated with different Markers</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_UncodedReference.rpt">Alleles with most uncoded references</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_NoPhenoRef.rpt">Alleles with no phenotype but have additional references</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_NoMCL.sql.rpt">Alleles of type Targeted or Gene Trap that have no MCL association</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_MCLAllele.sql.rpt">MCL's that are associated with more than one Allele</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_Notes.rpt">Allele Notes with HTML errors</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_OrphanGenotype.sql.rpt">Orphan Genotypes</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_NewAllele.current.rpt">New Allele Report</A> <B>(acv)</B>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_RefCounts.rpt">Number of Allele References By Year</A> <B>(csmith, steve-rockwood)</B>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_ExpressedComp.sql.rpt">Allele Missing Expressed Component Attribute</A> <B>(mnk)</B>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_WithdrawnMarker.sql.rpt">Alleles with Withdrawn Markers</A> <B>(mnk)</B>
<LI><A HREF="/data/reports/qcreports_db/archive/alleles">New Allele Archive</A>

</OL>

<A NAME="phenoannot"><B>Phenotype Vocabulary Annotation (csmith, smb, dlb, honda, mnk)</B> </A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_MPObsolete.sql.rpt">Genotype Annotations to Obsolete MP Terms</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_OMIMObsolete.sql.rpt">Genotype Annotations to Obsolete OMIM Terms</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_MPDups.sql.rpt">Duplicate MP Annotations</A>
<LI><A HREF="/data/loads/mgi/vocload/OMIM/OMIMtermcheck.current.rpt">OMIM term or ID changes</A>
<LI><A HREF="/data/loads/mgi/vocload/OMIM/OMIM.clusters.diff1">OMIM ids in MGI but not in OBO-Disease-Cluster file</A>
<LI><A HREF="/data/loads/mgi/vocload/OMIM/OMIM.clusters.diff2">OMIM ids in OBO_Disease-Cluster file but not in MGI</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_MPAnnot.rpt">Total MP Annotations per Gene</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_MPNewTerms.rpt">MP Ontology terms added in the last week</A>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_GenoNoMarker.rpt">Genotype Alleles that Contain No Markers</A>
</OL>

<A NAME="gxd"><B>GXD (cms, jfinger, ijm, terryh)</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_Triage.sql.current.rpt">GXD Triage (weekly)</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_Stats.current.rpt">GXD Statistics</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_StatsMonthly.rpt">GXD Monthly Statistics by Assay</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_NotInCache.sql.rpt">GXD Assays missing from GXD Expression Cache Table</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SpecTheilerAge.rpt">GXD Specimens with incompatible Theiler stages and ages</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SpecNoAge.sql.rpt">Specimens with incorrect age/age range values</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SameStructure.rpt">GXD Specimens/Lanes that use the same Structure > 1</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SpecTheiler.sql.rpt">Specimens annotated to structures of > 1 Theiler Stage or Sex</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ExpPresNotPres.rpt">Assays in which the same anatomical structure is annotated as having expression both present and absent</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ChildExpNotParent.rpt">Expression present in child but not parent</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ImagesUnused.rpt">Unused image panes from full coded references</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ImageCheck.sql.rpt">GXD Image Check</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ImageJPG.sql.rpt">Displayed Image Panes annotated to more than 2 assays</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ImageSpecimen.rpt">Image panes annotated to >1 specimens in same assay</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_FullCodeGene.sql.rpt">Full Coded References that contain indexed Genes that have not been full coded</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_FullCodeMissingIndex.sql.rpt">Full Coded References that are missing a Gene in the Index</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_AssayInCacheNotInIndex.rpt">Full coded assays that were not indexed</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_AssayInIndexNotInCache.rpt">Full coded references that contain indexed assays that have not been full coded</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_GenotypeRef.sql.rpt">References where GXD and Pheno differ in Genotype</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ProbeAntibody.sql.rpt">Problems with GXD Assay Probes/Antibodies</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_WithdrawnMarker.rpt">Withdrawn markers with GXD data</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_NonGeneMarkers.rpt">Non-gene markers with GXD annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_NoSequence.rpt">Genes without Sequence with GXD Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_GenotypeNA.sql.rpt">GXD Annotations to "Not Applicable" Genotype</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_GelBand.sql.rpt">GXD Blot w/ Control = No and Strength = Not Applicable</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_DupPrimers.sql.rpt">Duplicate Primer Sets used in GXD</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_DuplicateSpecimenLabel.sql.rpt">Duplicate Specimen Labels used in GXD</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_GenotypeHemi.sql.rpt">Hemizygous Genotype Check</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_KnockInGene.sql.rpt">GXD Knock Ins where the assayed Gene is not the Gene which is mutated in the Genotype</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SeqID.rpt">Probes used in GXD Assay w/ Seq IDs/Seq ID Attached to another Marker</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Reference.sql.rpt">Probes/Primers w/out References</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_Images.rpt">Papers Requiring Images</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_LackingImages.rpt">Papers Requiring Permissions</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ImagePanes.rpt">Panes Lacking Coordinates</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_FullCodeableIndexed.rpt">Indexed-Full-Coded gene list</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_FullCodeable2.rpt">Papers containing genes that are not in the full coded portion of the database</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_FullCodeable.rpt">Markers that have no full coded data</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_EarlyPapers.rpt">Early paper for popular genes</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_MechDev.rpt">Mech Dev papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_GeneExprPatternsandBrainResGeneExprPatterns.rpt">Gene Expr Patterns papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_DevBiol.rpt">Dev Biol papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_DevDyn.rpt">Dev Dyn papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_Development.rpt">Development papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_References.sql.rpt">Reference selected for GXD which have been not been used</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_EMAPA_Terms.rpt">EMAPA Terms</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_EMAPS_Terms.rpt">EMAPS Terms</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_CellLine.sql.rpt">Cell Lines</A>
<LI><A HREF="/data/reports/qcreports_db/archive/gxd">Archived GXD Reports</A>
</OL>

<A NAME="recomb"><B>Recombinase and Transgenic Reporter Data (mnk, cms)</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_NotInCache.sql.rpt">Recombinant/Transgenic Assays missing from Cache Table</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_SpecNoAge.sql.rpt">Recombinant/Transgenic Specimens with incorrect age/age range values</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_SpecTheiler.sql.rpt">Recombinant/Transgenic Specimens annotated to structures of > 1 Theiler Stage or Sex</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_SpecTheilerAge.rpt">Recombinant/Transgenic Specimens with incompatible Theiler stages and ages</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_GenotypeHemi.sql.rpt">Recombinant/Transgenic Hemizygous Genotype Check</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_SameStructure.rpt">Recombinant/Transgenic Specimens that use the same Structure > 1</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_ExpPresNotPres.rpt">Signal present and absent in the same structure in the same Recombinant/Transgenic Assay</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_ChildExpNotParent.rpt">Signal present in child but not parent in the same Recombination/Transgenic Assay</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_Images.rpt">Recombinase Papers Requiring Images</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_LackingImages.rpt">Recombinase Papers Requiring Permissions</A>
</OL>

<A NAME="go"><B>GO (hjd, dph, jblake, dsitnikov, ln, mdolan)</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_RatNoGO.sql.rpt">Mouse Genes that have Rat/Human Homologs but no GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_MRK_NoGO.rpt">Mouse Genes with no GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_MRK_NoGO_Has_Alleles.rpt">Mouse Genes that have Alleles but no GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoGO_D.html">Genes with no GO Annotations with references that are selected for GO but have not been used</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_AllNoGO.html">Genes with references that are selected for GO but have not been used</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOIEA_D.html">Genes with GO Annotations of evidence IEA only and with references that are selected for GO but have not been used</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOIEA_E.rpt">Genes with OMIM Annotations and either GO Annotations of evidence IEA only or no GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOGold_B.rpt">All Genes with Annotations to IDA, IGI, IMP, IPI</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOGold_D.rpt">All Genes with multiple Annotations to IDA, IGI, IMP, IPI</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_EvidenceCodeStats.sql.rpt">Total # of Annotations by GO Evidence Code</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_InvalidInferred.rpt">Invalid "Inferred From" Values in GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_InvalidProperties.rpt">Invalid "Properties" Values in GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_stats.rpt">GO Statistics</A>
<LI><A HREF="/data/reports/qcreports_db/archive/go">Archive of GO Statistics</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOUnknown.sql.rpt">All genes with 'unknown' annotations with new indexed literature</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_done.mgi">"Done" Genes with Literature</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_NotGene.sql.rpt">Non-Gene Markers with GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_GOObsolete.rpt">Markers with Annotations to Obsolete GO Terms</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GORCA.html">Genes with GO Annotations of evidence RCA only and with references that are selected for GO but have not been used</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_EvidenceProperty.rpt">J:73065 GO annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_eco_multipleEvidence.rpt">ECO : report</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_eco_multipleEvidence.log">ECO : log</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_Combined_Report.rpt">Combined QC Report</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOAnnot.rpt">Total GO Annotations per Gene</A> <B>(jblake)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_GenotypeDescriptions.rpt">Genotype Descriptions with MGI IDs</A> <B>(mdolan)</B>
</OL>

<A NAME="mtb"><B>MTB (dab, dmk)</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MTB_Triage.current.txt">MTB Triage (weekly)</A>
<LI><A HREF="/data/reports/qcreports_db/archive/mtb">Archived MTB Reports</A>
</OL>

<A NAME="nomenclature"><B>Nomenclature (mmh)</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/NOMEN_Broadcast.sql.rpt">Nomenclature Symbols Broadcast Within Last 3 Days</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_SymbolDiffs1.rpt">MGI Symbols differing from Human Ortholog Symbols, excludes RIKEN (#1)</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_SymbolDiffs3.rpt">MGI Symbols differing from Human Ortholog Symbols, RIKEN only (#1)</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Withdrawn.sql.rpt">Withdrawn Markers</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_InterimByDate.sql.rpt">Interim Markers (sorted by human symbol status/date)</A>
<LI><A HREF="/data/reports/qcreports_db/output/BIB_RefsToIndex.rpt">References to be indexed</A>
</OL>

<A NAME="strains"><B>Strains (brs)</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainStd.sql.rpt">Strains (Standard only)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainNonStd.sql.rpt">Strains (Non Standard excluding F1 and F2)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainF1F2.sql.rpt">Strains Containing F1 or F2</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainDup.sql.rpt">Duplicate Strains</A> <B>(dlb)</B>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainByDate.sql.rpt">Strains (Non Standard - Sorted by Creation Date)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain.rpt">Tab-delimited File of All Strains (MGI Acc ID, Strain, Markers, Synonyms, External Acc IDs)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain2.rpt">Tab-delimited File of All Strains (Strain/Standard)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainTrans.sql.rpt">Strains - Transgenic or Targeted Mutation</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainNonStdNoData.sql.rpt">Non-Standard Strains with no JR# and no data attached</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_MultNoAlleles.sql.rpt">Mutant Strains w/ Multiple Markers and No Alleles</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_SingleNoAlleles1.sql.rpt">Mutant Strains w/ Single Markers, No Alleles and Allele Symbol is in MGI</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_SingleNoAlleles2.sql.rpt">Mutant Strains w/ Single Markers, No Alleles and Allele Symbol is not in MGI</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_NoAlleles1.sql.rpt">Strains w/ Markers, No Alleles and Allele Symbol is in MGI</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLP_Nomenclature.sql.rpt">Strains Affected by recent Nomenclature changes</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_NomenSyns.rpt">Strain Nomenclature Synonyms</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_OtherSyns.rpt">Strain Other Synonyms</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain4.rpt">Strains with MGI IDs that have more than one additional external ID</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain5.rpt">Strains where any of their IDs contain an "O"</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain6.rpt">Strains containing "either" or "involves" and marked STANDARD</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_Standard.sql.rpt">STOCK or 'semicolon' strains marked standard</A>
</OL>

<A NAME="strainsjax"><B>Strains - JAX (acv)</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX2.jrs.current.rpt">Public JAX strains/Alleles used in Genotype/No Strain/Genotype Association</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX3.jrs.rpt">JAX Strains w/ Genotype Associations where the Markers/Alleles of the Strain record do not exactly match the Markers/Alleles of the Genotype record</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX4.sql.current.rpt">New JR#s created in the past week</A> <B>(dlb)</B>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX5.sql.current.rpt">JR#s with alleles added within the past week</A>
<LI><A HREF="/data/reports/qcreports_db/archive/strain">Strain Archive</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX2.mmrrc.current.rpt">Public MMRRC strains/Alleles used in Genotype/No Strain/Genotype Association</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX3.mmrrc.rpt">MMRRC Strains w/ Genotype Associations where the Markers/Alleles of the Strain record do not exactly match the Markers/Alleles of the Genotype record</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX7.sql.current.rpt">New MMRRC JR#s created in the past week</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX8.sql.current.rpt">MMRRC JR#s with alleles added within the past week</A>
</OL>

<A NAME="sequencegroup"><B>Sequence Group (rmb, cjb, djr, yz)</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_NMs.rpt">NM Sequences that are not associated with a MGI Marker</A> <B>(yz, djr)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_GenesAndPseudogenesWithSequence.rpt">Genes & Pseudogenes with at least one Sequence association</A> <B>(?)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_ENSEMBL_Associations.rpt">Ensembl Associations</A> <B>(?)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_VEGA_Associations.rpt">VEGA Associations</A> <B>(?)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_NCBI_Associations.rpt">NCBI Associations</A> <B>(?)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GmNoGeneModel.sql.rpt">Gm Markers without Gene Model Associations</A> <B>(yz, djr)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_ObsoleteGeneModel.sql.rpt">Obsolete Gene Models associated with Markers</A> <B>(yz, djr)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GeneNoMcvAnnot.rpt">Genes with no MCV Annotation</A> <B>(yz)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_McvAnnotByFeature.rpt">MCV Annotations by Feature</A> <B>(djr)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_McvAnnot.rpt">Marker MCV Annotations</A> <B>(yz, djr)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/mcvload/index.html">MCV Annotation Load - these reports are refreshed only when a batch annotation load is run</A> <B>(djr)</B>
<LI><A HREF="/data/loads/mgi/mrkcacheload/logs/mrkmcv.curator.log">MCV Cache Load - this report is refreshed nightly M-F</A>  <B>(djr)</B>
<BR>This report contains the following sections:
    <OL>
    <LI>Markers  with conflict between the Marker Type and the MCV Marker Type
    <LI>Markers Annotated to Grouping Terms
    <LI>Markers with Multiple MCV Annotations
    </OL>
<LI>The following logs report: <B>(djr)</B>
    <BR>* Gene Models in MGI that have been obsoleted by the provider 
    <BR> * Gene Model biotypes not loaded because unable to translate biotype. 
    <OL>
    <LI><A HREF="/data/loads/mgi/genemodelload/logs/ensembl_seqgenemodelload.log"> Ensembl Sequence Gene Model Information Load</A>
    <LI><A HREF="/data/loads/mgi/genemodelload/logs/vega_seqgenemodelload.log"> VEGA Sequence Gene Model Information Load</A>
    <LI><A HREF="/data/loads/mgi/genemodelload/logs/ncbi_seqgenemodelload.log"> NCBI Sequence Gene Model Information Load</A>
    </OL>
<LI><A HREF="/data/reports/qcreports_db/output/GIA_MRK_ENSEMBL.rpt">Ensembl Associations for Gene Index</A> <B>(yz)</B>
<LI><A HREF="/data/reports/qcreports_db/output/GIA_MRK_Sequence.rpt">Sequence Associations for Gene Index</A> <B>(yz)</B>
<LI><A HREF="/data/reports/reports_db/output/MGI_BioTypeConflict.rpt">MGI Biotype Conflict Public Report</A> <B>(cjb)</B>
<LI>QTL Reports
    <OL>
    <LI><A HREF="/data/reports/qcreports_db/output/MRK_QTL_1.rpt">QTL markers that have no mapping and no allele associations</A> <B>(djr)</B>
    <LI><A HREF="/data/reports/qcreports_db/output/MRK_QTL_2.rpt">QTL markers that have mapping but no allele associations</A> <B>(djr)</B>
    <LI><A HREF="/data/reports/qcreports_db/output/MRK_QTL_3.rpt">QTL markers that have no mapping but have allele associations</A> <B>(djr)</B>
    <LI><A HREF="/data/reports/qcreports_db/output/MRK_QTL_4.rpt">QTL markers that have assigned reference of J:23000 or J:85000</A> <B>(djr)</B>
    <LI><A HREF="/data/reports/qcreports_db/output/MRK_QTL_5.rpt">QTL markers that are in Nomen and have not been broadcast</A> <B>(djr)</B>
    <LI><A HREF="/data/reports/qcreports_db/output/QTL_Triage.rpt">Master Bib QTL Indexing Report</A> <B>(djr)</B>
    <LI><A HREF="/data/reports/qcreports_db/output/MGI_QTL_ReferenceNotes.sql.rpt">QTL Public Reference Notes</A> <B>(djr)</B>
    </OL>
</OL>

<A NAME="sequence"><B>Sequences (rmb, cjb, djr, yz)</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_Deleted.sql.rpt">Deleted Sequences with MGI Associations</A>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_Split.sql.rpt">Split Sequences</A>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_MarkerAnnotation.sql.rpt">Markers Annotated to a Secondary Sequence Accession ID</A>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_Dummy.sql.rpt">Dummy Sequences Annotated to Mouse Marker (report a) & Molecular Segments (report b) </A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_MultCoordinate.sql.rpt">Symbols w/ > 1 Ensembl, VEGA or NCBI Gene Model Association</A> <B>(yz,djr)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_MultMarkerGeneModels.sql.rpt">Ensembl, VEGA or NCBI Gene Models w/ > 1 Marker Association</A> <B>(yz,djr)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoENSEMBL.rpt">Ensembl Gene Models with no Marker Association</A> <B>(yz,djr)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoVEGA.rpt">VEGA Gene Models with no Marker Association</A> <B>(yz,djr)</B>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoNCBI.rpt">NCBI Gene Models with no Marker Association</A> <B>(yz,djr)</B>
</OL>

<A NAME="homology"><B>Homology (jblake, mdolan) </B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Homology_Class1.rpt">Classes for which there are > 10 Genes From a Single Species</A> 
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Homology_Class2.rpt">Classes with no Mouse Genes</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Homology_Class3.rpt">Classes with only Mouse Genes</A>
</OL>
<A NAME="dataloads"><B>Data Loads</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/live/dataload/egload/index.html">EntrezGene Load Mouse</A> <B>(yz,djr)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/entrezgeneload/index.html">EntrezGene Load Human/Rat/etc</A>
<LI><A HREF="/usrlocalmgi/live/dataload/uniprotload">UniProt Load Data and Logs</A>
<LI><A HREF="/usrlocalmgi/live/dataload/unigeneload">UniGene Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/genemodelload/index.html">Gene Model/Association Loads</A>
<LI><A HREF="/usrlocalmgi/live/dataload/mousecycload/index.html">MouseCyc Load (vocabulary, annotations)</A>
<LI><A HREF="/usrlocalmgi/live/dataload/pirsfload/index.html">PIRSF Load</A> <B>(jblake)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/ccdsload/index.html">CCDS Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/vocload/MAindex.html">Adult Mouse Anatomy Vocabulary/DAG Load</A> <B>(terryh)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/emageload/index.html">EMAGE Load</A> <B>(cms)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/geoload/index.html">GEO Load</A> <B>(cms)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/gensatload/index.html">GENSAT Load</A> <B>(cms)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/arrayexpload/index.html">ArrayExpress Load</A> <B>(cms)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/vocload/GOindex.html">GO Vocabulary/DAG Load</A> <B>(hjd)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/goaload/index.html">GOA/Mouse Load</A> <B>(hjd)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/goahumanload/index.html">GOA/Human Load</A> <B>(hjd)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/goratload/index.html">GO/Rat Load</A> <B>(hjd)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/gorefgenload/index.html">GO/PAINT Load</A> <B>(hjd)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/gocfpload/index.html">GO/CFP (aka GOC) Load</A> <B>(hjd)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/homologyload/index.html">Homology Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/vocload/MPindex.html">Mammalian Phenotype Vocabulary/DAG Load</A> <B>(csmith)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/vocload/OMIMindex.html">OMIM Vocabulary Load</A> <B>(smb)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/vocload/CLindex.html">Cell Ontology (CL) Vocabulary Load</A> <B>(dph)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/genmapload/index.html">Genetic Map Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/mcvload/index.html">MCV Annotation Load</A>
<LI><A HREF="/data/loads/mgi/mrkcacheload/logs/mrkmcv.curator.log">MCV Cache Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/alomrkload/index.html">ALO Marker Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/htmpload/impc_index.html">HTMP load: IMPC</A> <B>(csmith, cms)</B>
<LI><A HREF="/usrlocalmgi/live/dataload/targetedalleleload/index.html">Targeted Allele Load</A> <B>(csmith)</B>
</OL>

<HR>

</BODY>
</HTML>
