<HTML>
<HEAD>
<TITLE>
Quality Control Reports
</TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF">

<HR NOSHADE>
<H1>MGI Quality Control Reports</H1>
<HR>
QC reports are generated nightly from the MGD Production Database (unless otherwise noted).
They are intended for internal use only (i.e. not available for public consumption, although they may
be used at times to test reports for <I>future</I> public consumption.)
<HR>

<B>Table of Contents</B>
<UL>
<LI><A HREF="#references">References</A>
<LI><A HREF="#markers">Markers</A>
<LI><A HREF="#alleles">Alleles</A>
<LI><A HREF="#phenoannot">Phenotype Vocabulary Annotation</A>
<LI><A HREF="#probes">Probes</A>
<LI><A HREF="#mapping">Mapping</A>
<LI><A HREF="#orthology">Orthology</A>
<LI><A HREF="#gxd">GXD</A>
<LI><A HREF="#recomb">Recombinase and Transgenic Reporter Data</A>
<LI><A HREF="#go">GO</A>
<LI><A HREF="#mtb">MTB</A>
<LI><A HREF="#nomenclature">Nomenclature</A>
<LI><A HREF="#strains">Strains</A>
<LI><A HREF="#strainsjax">Strains - JAX</A>
<LI><A HREF="#sequencegroup">Sequence Group</A>
<LI><A HREF="#sequence">Sequence</A>
<LI><A HREF="#dataloads">Data Loads</A>
<LI><A HREF="#lists">Lists/Other</A>
<LI><A HREF="ftp://ftp.informatics.jax.org/pub/reports/index.html">Public Reports</A>
</UL>
<HR>

<A NAME="references"><B>References</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/BIB_NeedNLM1.sql.rpt">References Which Need NLM Updates</A>
<LI><A HREF="/data/reports/qcreports_db/output/BIB_InPress.sql.rpt">References In Press/Submissions</A>
<LI><A HREF="/data/reports/qcreports_db/output/BIB_NoAbstract.sql.rpt">References w/ PubMed ID and No Abstract</A>
<LI><A HREF="/data/reports/qcreports_db/output/BIB_NoRef.sql.rpt">References with no Marker Attached</A>
<LI><A HREF="/data/reports/qcreports_db/output/BIB_RefsToIndex.rpt">References to be indexed</A>
</OL>

<A NAME="markers"><B>Markers</B> </A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_OffsetsMLDP.sql.rpt">Markers with No Offsets Modifed in Mapping w/in the last 3 days</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Offsets.sql.rpt">Markers w/ MGD Offset but w/out CC Offset or Mapping Information</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_OffsetsUN.sql.rpt">Markers with Offsets but no Chromosome Assignment</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Matrix.sql.rpt">UN/Syntenic Markers In MGD found in Mapping Panel Crosses</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_LowerSeq.rpt">Lowercase Gene Symbols which have GenBank IDS</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_AssigningRef.rpt">MGI Markers with Assigning Reference</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_MGC.sql.rpt">MGC IDs with Nomen Records Not Deleted or Broadcast</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoSequence.rpt">Genes with No Sequence Data, Excluding Mutant Phenotypes</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_QTL.rpt">QTL Markers added to MGI since March 2002 with no MP annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_SharedSynonyms.rpt">Markers with Shared Synonyms</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_TranscriptNoSP.sql.rpt">MGI genes that have Transcript Sequences but no SwissProt association</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_New_QTLS.rpt">QTL Markers since the previous week</A>
</OL>

<A NAME="alleles"><B>Alleles</B> </A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Alleles.sql.rpt">Allele Anomalies</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_Progress.current.rpt">Allele Progress Report</A>
<LI><A HREF="/data/reports/qcreports_db/archive/alleles">Allele Progress Archive</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_NoMutantSentence.rpt">Genes w/Phenotypic Alleles and No Mutant Sentence</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_MarkerClip.sql.current.rpt">Genes with Marker Clips and Alleles that have been added in the last 5 days</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_DupGenotype.rpt">Duplicate Genotypes</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_NoMolMutation.rpt">Alleles that have Molecular Notes but no Molecular Mutation</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_MolNotesNoMP.rpt">Alleles that have Molecular Notes but no MP Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_GeneTrap.rpt">Gene Trap Allele Lab Codes and their Max Sequence Number</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_WildType.sql.rpt">Genotypes where Allele 1 is "wild type"</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_Priority.rpt">Alleles that have 5 or more Priority References</A>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_CellLineProvider.sql.rpt">Mutant ES Cell Line Provider Check</A>
<LI><A HREF="/data/reports/qcreports_db/output/PHENO_ImageCheck.sql.rpt">Image Check</A>
<LI><A HREF="http://aretha.jax.org/jax/tbase">TBASE</A>
</OL>

<A NAME="phenoannot"><B>Phenotype Vocabulary Annotation</B> </A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/ALL_ImmuneAnnot.current.rpt">Alleles of Genotype Annotations with Header Term: immune system</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_MPObsolete.sql.rpt">Genotype Annotations to Obsolete MP Terms</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_OMIMObsolete.sql.rpt">Genotype Annotations to Obsolete OMIM Terms</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_MPDups.sql.rpt">Duplicate MP Annotations</A>
<LI><A HREF="/data/loads/mgi/vocload/OMIM/OMIMtermcheck.rpt">OMIM term or ID changes</A>
</OL>

<A NAME="probes"><B>Probes</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_NoGenBank.sql.rpt">No GenBank IDs - Sequence in Paper</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_NoMarkerSeq.sql.rpt">Molecular Segments with Encoding Markers and Sequence IDs and no corresponding Marker-Sequence ID association</A>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_ProblemSequence.rpt">Molecular Segment with the Problem Sequence Note associated with more than one GenBank Seqid or associated with no GenBank Seqid</A>
</OL>

<A NAME="mapping"><B>Mapping</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MLD_Triage.sql.current.rpt">Mapping Triage (weekly)</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLD_Chromosome.sql.rpt">Invalid Chromosomes</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLD_Cross.sql.rpt">Invalid Cross Data</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLD_Marker.sql.rpt">Missing Experimental Markers</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLD_DupMarker.sql.rpt">Duplicate Markers Appearing in Experiments</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLD_NomenText.sql.rpt">Symbols which have undergone withdrawals and appear in Text Experiments</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_MappedNoRef.sql.rpt">Mapped Markers w/out a Mapping Reference</A>
<LI><A HREF="/data/reports/qcreports_db/archive/mapping">Archived Mapping Reports</A>
</OL>

<A NAME="orthology"><B>Orthology</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_References2.sql.rpt">References selected for Orthology only w/out Orthology Data</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_References5.sql.rpt">References selected for Orthology w/ no Orthology Data with at least one reference-to-gene Association</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_SpeciesCount.sql.rpt">Counts of Mouse Orthologies by Organism</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_Duplicates.sql.rpt">Duplicate Orthologies (by Class, Reference, Species)</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_ConservedUN.sql.rpt">Orthologies w/ Conserved Map Location and Chromosome = UN</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_NoMouseSeq.sql.rpt">Mouse Markers in Orthology w/ Nucleotide or Amino Acid Sequence Similarity Assay and Marker has no Seq ID</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_MouseSeq1.rpt">Mouse Genes with Sequence ID but no Human Orthology</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_MouseSeq2.sql.rpt">Mouse Genes with no Sequence ID but with Human Orthology</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_NoMouse.sql.rpt">Orthology Records w/out Mouse Genes</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_HomoloGene1.sql.rpt">HomoloGene Mouse/Human Orthologous Pairs where we have neither the mouse nor the human member of the pair. (HUMAN AND MOUSE SYMBOLS ARE THE SAME)</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_HomoloGene2A.sql.rpt">HomoloGene Mouse/Human Orthologous Pairs where we have neither the mouse nor the human member of the pair. (HUMAN AND MOUSE SYMBOLS ARE NOT THE SAME, HUMAN SYMBOLS ARE OFFICIAL)</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_HomoloGene2B.sql.rpt">HomoloGene Mouse/Human Orthologous Pairs where we have neither the mouse nor the human member of the pair. (HUMAN AND MOUSE SYMBOLS ARE NOT THE SAME, HUMAN SYMBOLS ARE INTERIM)</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_HomoloGene3.sql.rpt">HomoloGene Mouse/Human Orthologous Pairs where our Orthology contains Human but no Mouse member.</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_HomoloGene4.sql.rpt">HomoloGene Mouse/Human Orthologous Pairs where our Orthology contains Mouse but no Human member.</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_Singletons.sql.rpt">Mouse/Human Singletons</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_RatNoSeq.sql.rpt">Rat Genes with no Sequence ID but with Orthology</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NomenRat.sql.rpt">Mouse Symbols with nomenclature changes which contain Rat homologs with unofficial nomenclature</A>
</OL>

<A NAME="gxd"><B>GXD</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_Triage.sql.current.rpt">GXD Triage (weekly)</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ProbeAntibody.sql.rpt">Problems with GXD Assay Probes/Antibodies</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_NotInCache.sql.rpt">GXD Assays missing from GXD Expression Cache Table</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_WithdrawnMarker.rpt">Withdrawn markers with GXD data</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_NonGeneMarkers.rpt">Non-gene markers with GXD annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_Stats.current.rpt">GXD Statistics</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_StatsMonthly.rpt">GXD Monthly Statistics by Assay</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SpecNoAge.sql.rpt">Specimens with incorrect age/age range values</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SpecTheiler.sql.rpt">Specimens annotated to structures of > 1 Theiler Stage or Sex</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SpecTheilerAge.rpt">GXD Specimens with incompatible Theiler stages and ages</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_GenotypeHemi.sql.rpt">Hemizygous Genotype Check</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_MechDev.rpt">Mech Dev papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_GeneExprPatternsandBrainResGeneExprPatterns.rpt">Gene Expr Patterns papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_DevBiolandDevelopment.rpt">Dev Biol and Development papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_DevDyn.rpt">Dev Dyn papers to full code</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_FullCodeable.rpt">Markers that have not been full coded and that have full codeable papers</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_FullCodeable2.rpt">Papers containing genes that are not in the full coded portion of the database</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_EarlyPapers.rpt">Early paper for popular genes</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_EarlyEmbPapers.rpt">Early Embryonic Expression Papers</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_FullCodeable3.rpt">Blot-only papers that have not been full coded</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_OneGenePapers.rpt">One gene papers</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_FullCodeGene.sql.rpt">Full Coded References that contain indexed Genes that have not been full coded</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_FullCodeMissingIndex.sql.rpt">Full Coded References that are missing a Gene in the Index</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_KnockInGene.sql.rpt">GXD Knock Ins where the assayed Gene is not the Gene which is mutated in the Genotype</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_GenotypeRef.sql.rpt">References where GXD and Pheno differ in Genotype</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_GenotypeNA.sql.rpt">GXD Annotations to "Not Applicable" Genotype</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_Images.rpt">Papers Requiring Images</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_LackingImages.rpt">Papers Requiring Permissions</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ImageCheck.sql.rpt">GXD Image Check</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SameStructure.rpt">GXD Specimens/Lanes that use the same Structure > 1</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_GelBand.sql.rpt">GXD Blot w/ Control = No and Strength = Not Applicable</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_DupPrimers.sql.rpt">Duplicate Primer Sets used in GXD</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ImageJPG.sql.rpt">Displayed Image Panes annotated to more than 2 assays</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ImagesUnused.rpt">Unused image panes from full coded references</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_SeqID.rpt">Probes used in GXD Assay w/ Seq IDs/Seq ID Attached to another Marker</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ExpPresNotPres.rpt">Assays in which the same anatomical structure is annotated as having expression both present and absent</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_ChildExpNotParent.rpt">Expression present in child but not parent</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Reference.sql.rpt">Probes/Primers w/out References</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Marker.sql.rpt">a) Duplicate Markers b) No Markers (excluding MGC, IMAGE, RPCI, NIA, RIKEN clones)</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_RefNoTitle.rpt">GXD references lacking titles</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_Transgenic.rpt">Transgenic Genotype Check of the Expression Cache</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_AssayInCacheNotInIndex.rpt">Full coded assays that were not indexed</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_AssayInIndexNotInCache.rpt">Full coded references that contain indexed assays that have not been full coded</A>
<LI><A HREF="/data/reports/qcreports_db/archive/gxd">Archived GXD Reports</A>
</OL>

<A NAME="recomb"><B>Recombinase and Transgenic Reporter Data</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_NotInCache.sql.rpt">Recombinant/Transgenic Assays missing from Cache Table</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_SpecNoAge.sql.rpt">Recombinant/Transgenic Specimens with incorrect age/age range values</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_SpecTheiler.sql.rpt">Recombinant/Transgenic Specimens annotated to structures of > 1 Theiler Stage or Sex</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_SpecTheilerAge.rpt">Recombinant/Transgenic Specimens with incompatible Theiler stages and ages</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_GenotypeHemi.sql.rpt">Recombinant/Transgenic Hemizygous Genotype Check</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_SameStructure.rpt">Recombinant/Transgenic Specimens that use the same Structure > 1</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_ExpPresNotPres.rpt">Signal present and absent in the same structure in the same Recombinant/Transgenic Assay</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_ChildExpNotParent.rpt">Signal present in child but not parent in the same Recombination/Transgenic Assay</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_ImageCheck.sql.rpt">Recombinase Image Check</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_Images.rpt">Recombinase Papers Requiring Images</A>
<LI><A HREF="/data/reports/qcreports_db/output/RECOMB_LackingImages.rpt">Recombinase Papers Requiring Permissions</A>
</OL>

<A NAME="go"><B>GO</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_RatNoGO.sql.rpt">Mouse Genes that have Rat/Human Homologs but no GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoGO_A.rpt">Genes with no GO Annotations and reference count excludes data loads and GenBank submissions</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoGO_B.rpt">Genes with no GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoGO_C.rpt">Genes with no GO Annotations and reference count includes only data loads and GenBank submissions</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoGO_D.html">Genes with no GO Annotations with references that are selected for GO but have not been used</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_AllNoGO.html">Genes with references that are selected for GO but have not been used</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOIEA_A.rpt">Genes with GO Annotations of evidence IEA only and reference count excludes data loads and GenBank submissions</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOIEA_C.rpt">Genes with GO Annotations of evidence IEA only and reference count includes only data loads and GenBank submissions</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOIEA_D.html">Genes with GO Annotations of evidence IEA only and with references that are selected for GO but have not been used</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOIEA_E.rpt">Genes with OMIM Annotations and either GO Annotations of evidence IEA only or no GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOIEA_F.rpt">Genes with Alleles and either GO Annotations of evidence IEA only or no GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOGold_A.rpt">All Genes with Annotations to IDA, IGI, IMP, IPI, TAS</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOGold_B.rpt">All Genes with Annotations to IDA, IGI, IMP, IPI</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOGold_C.rpt">All Genes with multiple Annotations to IDA, IGI, IMP, IPI, TAS</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOGold_D.rpt">All Genes with multiple Annotations to IDA, IGI, IMP, IPI</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOAllele.rpt">Genes with Annotation to IMP, Inferred From = NULL and Reference Associations to Alleles</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_EvidenceCodeStats.sql.rpt">Total # of Annotations by GO Evidence Code</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_InvalidInferred.rpt">Invalid "Inferred From" Values in GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_stats.rpt">GO Statistics</A>
<LI><A HREF="/data/reports/qcreports_db/archive/go">Archive of GO Statistics</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GOUnknown.sql.rpt">All genes with 'unknown' annotations with new indexed literature</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_GODups.sql.rpt">Duplicate GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/gp2protein_SWP_TR.mgi">Special gp2protein_SWP_TR.mgi (includes TrEMBL)</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_done.mgi">"Done" Genes with Literature</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_NotGene.sql.rpt">Non-Gene Markers with GO Annotations</A>
<LI><A HREF="/data/reports/qcreports_db/output/VOC_GOObsolete.rpt">Markers with Annotations to Obsolete GO Terms</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_GORCA.html">Genes with GO Annotations of evidence RCA only and with references that are selected for GO but have not been used</A>
<LI><A HREF="/data/reports/qcreports_db/output/GO_Combined_Report.rpt">Combined QC Report</A>
</OL>

<A NAME="mtb"><B>MTB</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MTB_Triage.current.txt">MTB Triage (weekly)</A>
<LI><A HREF="/data/reports/qcreports_db/archive/mtb">Archived MTB Reports</A>
</OL>

<A NAME="nomenclature"><B>Nomenclature</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/NOM_Triage.sql.current.rpt">Nomenclature Triage (weekly)</A>
<LI><A HREF="/data/reports/qcreports_db/output/NOMEN_Broadcast.sql.rpt">Nomenclature Symbols Broadcast Within Last 3 Days</A>
<LI><A HREF="/data/reports/qcreports_db/output/NOMEN_Pending.sql.rpt">Nomenclature In Progress Symbols</A>
<LI><A HREF="/data/reports/qcreports_db/output/NOMEN_Reserved.rpt">Nomenclature Reserved</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_SymbolDiffs1.rpt">MGI Symbols differing from Human Ortholog Symbols, excludes RIKEN (#1)</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_SymbolDiffs2.rpt">MGI Symbols differing from Human Ortholog Symbols, excludes RIKEN (#2)</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_SymbolDiffs3.rpt">MGI Symbols differing from Human Ortholog Symbols, RIKEN only (#1)</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_SymbolDiffs4.rpt">MGI Symbols differing from Human Ortholog Symbols, RIKEN only (#2)</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Withdrawn.sql.rpt">Withdrawn Markers</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_InterimByDate.sql.rpt">Interim Markers (sorted by Date)</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_ESTSynonyms.sql.rpt">EST IDs and Gm Markers as Synonyms</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NomenNotInformative.sql.rpt">MGI Markers with Uninformative Nomenclature #1</A>
<LI><A HREF="/data/reports/qcreports_db/output/HMD_RIKEN.sql.rpt">Human/Mouse Orthologies involving RIKEN Symbols</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Synonym_J94790.sql.rpt">Synonyms that require proper J number assignment</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_Synonym_Verify.sql.rpt">Synonyms that need to be verified</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_PirsfPseudogene.sql.rpt">Pseudogenes mapped to PIRSF superfamilies</A>
<LI><A HREF="/data/reports/qcreports_db/output/NOMEN_Uninformative_FAM_Markers.rpt">FAM orthologs with uninformative nomenclature</A>
<LI><A HREF="/data/reports/qcreports_db/archive/nomen">Archived Nomenclature Reports</A>
</OL>

<A NAME="strains"><B>Strains</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainStd.sql.rpt">Strains (Standard only)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainNonStd.sql.rpt">Strains (Non Standard excluding F1 and F2)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainF1F2.sql.rpt">Strains Containing F1 or F2</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainDup.sql.rpt">Duplicate Strains</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainByDate.sql.rpt">Strains (Non Standard - Sorted by Creation Date)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_NameChanges.rpt">Strain Names Which Have Been Changed</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain.rpt">Tab-delimited File of All Strains (MGI Acc ID, Strain, Markers, Synonyms, External Acc IDs)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain2.rpt">Tab-delimited File of All Strains (Strain/Standard)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainTrans.sql.rpt">Strains - Transgenic or Targeted Mutation</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainNonStdNoData.sql.rpt">Non-Standard Strains with no JR# and no data attached</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_MultNoAlleles.sql.rpt">Mutant Strains w/ Multiple Markers and No Alleles</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_SingleNoAlleles1.sql.rpt">Mutant Strains w/ Single Markers, No Alleles and Allele Symbol is in MGI</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_SingleNoAlleles2.sql.rpt">Mutant Strains w/ Single Markers, No Alleles and Allele Symbol is not in MGI</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_NoAlleles1.sql.rpt">Strains w/ Markers, No Alleles and Allele Symbol is in MGI</A>
<LI><A HREF="/data/reports/qcreports_db/output/MLP_Nomenclature.sql.rpt">Strains Affected by recent Nomenclature changes</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_PrivateRef.sql.rpt">Private Strains Referenced by Other Database Objects</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_NomenSyns.rpt">Strain Nomenclature Synonyms</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_OtherSyns.rpt">Strain Other Synonyms</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain3.rpt">Public Non-Standard Strains that have a MGI ID or an External ID</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain4.rpt">Strains with MGI IDs that have more than one additional external ID</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain5.rpt">Strains where any of their IDs contain an "O"</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain6.rpt">Strains containing "either" or "involves" and marked STANDARD</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Strain_Standard.sql.rpt">STOCK or 'semicolon' strains marked standard</A>
</OL>

<A NAME="strainsjax"><B>Strains - JAX</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX1.rpt">Public JAX Strains with no Annotation</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX2.current.rpt">Public JAX strains/Alleles used in Genotype/No Strain/Genotype Association</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX3.rpt">JAX Strains w/ Genotype Associations where the Markers/Alleles of the Strain record do not exactly match the Markers/Alleles of the Genotype record</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX4.sql.current.rpt">New JR##s created in the past week</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX5.sql.current.rpt">JR##s with alleles added within the past week</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_StrainJAX6.sql.rpt">Public JAX strains with nomen alleles but no genotype record</A>
<LI><A HREF="/data/reports/qcreports_db/archive/strain">Strain Archive</A>
</OL>

<A NAME="sequencegroup"><B>Sequence Group</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_ComprehensiveMarkerSequence.rpt">Comprehensive Marker/Sequence (bobs)</A>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_NMs.rpt">NM Sequences that are not associated with a MGI Marker</A>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_GenesAndPseudogenesWithSequence.rpt">Genes & Pseudogenes with at least one Sequence association</A>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_ENSEMBL_Associations.rpt">Ensembl Associations</A>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_VEGA_Associations.rpt">VEGA Associations</A>
<LI><A HREF="/data/reports/qcreports_db/output/MGI_NCBI_Associations.rpt">NCBI Associations</A>
</OL>

<A NAME="sequence"><B>Sequences</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_Deleted.sql.rpt">Deleted Sequences with MGI Associations</A>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_Split.sql.rpt">Split Sequences</A>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_MarkerAnnotation.sql.rpt">Markers Annotated to a Secondary Sequence Accession ID</A>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_Dummy.sql.rpt">Dummy Sequences Annotated to Mouse Marker (report a) & Molecular Segments (report b) </A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_MultCoordinate.sql.rpt">Symbols w/ > 1 Ensembl, VEGA or NCBI Gene Model Association</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_MultMarkerGeneModels.sql.rpt">Ensembl, VEGA or NCBI Gene Models w/ > 1 Marker Association</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_ENSEMBLnoGeneModel.rpt">Ensembl Genes with no Ensembl Gene Model Association</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_VEGAnoGeneModel.rpt">VEGA Genes with no VEGA Gene Model Association</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NCBInoGeneModel.rpt">NCBI Genes with no NCBI Gene Model Association</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_AllNoGeneModel.rpt">Ensembl/VEGA/NCBI Genes with Gene Model Association to different providers</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoENSEMBL.rpt">Ensembl Gene Models with no Marker Association</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoVEGA.rpt">VEGA Gene Models with no Marker Association</A>
<LI><A HREF="/data/reports/qcreports_db/output/MRK_NoNCBI.rpt">NCBI Gene Models with no Marker Association</A>
<LI><A HREF="/data/reports/qcreports_db/output/SEQ_RepTransGenomic.rpt">Representative Transcript and Genomic Sequences (bobs)</A>
</OL>

<A NAME="dataloads"><B>Data Loads</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/live/dataload/egload/index.html">EntrezGene Load Mouse: Data and QC Reports</A>
<LI><A HREF="/usrlocalmgi/live/dataload/entrezgeneload/index.html">EntrezGene Load Human/Rat/Chimp/Dog: Data and QC Reports</A>
<LI><A HREF="/usrlocalmgi/live/dataload/swissload">SwissProt Load Data and Logs</A>
<LI>UniGene</LI>
	<OL>
	<LI><A HREF="/data/reports/qcreports_db/output/multipleClusters.rpt">MGI Genes associated with multiple UniGene Clusters</A>
	<LI><A HREF="/data/reports/qcreports_db/output/multipleGenes.rpt">UniGene Clusters associated with multiple MGI Genes</A>
	<LI><A HREF="/data/reports/qcreports_db/output/genesMissingUG.rpt">MGI mouse Genes with Seq. ID but not in a UniGene Cluster</A>
	</OL>
<LI><A HREF="/usrlocalmgi/live/dataload/crossload/crossloadindex.html">CrossLoad: Data and QC Reports</A>
<LI><A HREF="/usrlocalmgi/live/dataload/vocload/GOindex.html">GO Vocabulary/DAG Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/goaload/index.html">GOA Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/vocload/MPindex.html">Mammalian Phenotype Vocabulary/DAG Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/vocload/MAindex.html">Adult Mouse Anatomy Vocabulary/DAG Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/vocload/OMIMindex.html">OMIM Vocabulary Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/mousecycload/index.html">MouseCyc Load (vocabulary, annotations)</A>
<LI><A HREF="/usrlocalmgi/live/dataload/mousefuncload/index.html">MouseFunc Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/orthologyload/index.html">Orthology Loads</A>
<LI><A HREF="/usrlocalmgi/live/dataload/emageload/index.html">EMAGE Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/geoload/index.html">GEO Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/gensatload/index.html">GENSAT Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/mgcload/index.html">MGC Association Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/rpciload/index.html">RPCI Clone Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/gbcdnaload/index.html">GenBank EST cDNA Clone Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/gbseqload/index.html">GenBank Sequence Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/refseqload/index.html">RefSeq Sequence Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/spseqload/index.html">SwissProt/TrEMBL Sequence Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/dfciseqload/index.html">DFCI Sequence Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/dotsseqload/index.html">DoTS Sequence Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/niaseqload/index.html">NIA Sequence Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/genemodelload/index.html">Gene Model/Association Loads</A>
<LI><A HREF="/usrlocalmgi/live/dataload/machipload/index.html">Microarray Association Loads</A>
<LI><A HREF="/usrlocalmgi/live/dataload/pirsfload/index.html">PIRSF Load</A>
<LI><A HREF="/usrlocalmgi/live/dataload/imsrload/index.html">IMSR Load</A>
</OL>


<A NAME="lists"><B>Lists/Other</B></A>
<OL>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_Tissue.sql.rpt">Tissues (All)</A>
<LI><A HREF="/data/reports/qcreports_db/output/PRB_CellLine.sql.rpt">Cell Lines</A>
<LI><A HREF="/data/reports/qcreports_db/output/ACC_ActualDB.sql.rpt">Logical/Actual DBs</A>
<LI><A HREF="/data/reports/qcreports_db/output/ACC_Accession.sql.rpt">Invalid EntrezGene/RefSeq IDs</A>
<LI><A HREF="/data/reports/qcreports_db/output/GXD_OrphanGenotype.sql.rpt">Orphan Genotypes</A>
</OL>

<HR>

</BODY>
</HTML>
