<HTML>
<HEAD>
<TITLE>
Quality Control Reports
</TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF">

<HR NOSHADE>
<H1>MGI Quality Control Reports</H1>
<HR>
QC reports are generated nightly from the MGD Production Database (unless otherwise noted).
They are intended for internal use only (i.e. not available for public consumption, although they may 
be used at times to test reports for <I>future</I> public consumption.)
<HR>

<B>Table of Contents</B>
<UL>
<LI><A HREF="#references">References</A>
<LI><A HREF="#markers">Markers</A>
<LI><A HREF="#alleles">Alleles</A>
<LI><A HREF="#mlc">MLC</A>
<LI><A HREF="#probes">Probes</A>
<LI><A HREF="#mapping">Mapping</A>
<LI><A HREF="#homology">Homology</A>
<LI><A HREF="#gxd">GXD</A>
<LI><A HREF="#go">GO</A>
<LI><A HREF="#fantom">Fantom2</A>
<LI><A HREF="#nomenclature">Nomenclature</A>
<LI><A HREF="#strains">Strains</A>
<LI><A HREF="#dataloads">Data Loads</A>
<LI><A HREF="#lists">Lists</A>
<LI><A HREF="#other">Other</A>
<LI><A HREF="ftp://ftp.informatics.jax.org/pub/reports/index.html">Public Reports</A>
</UL>
<HR>

<A NAME="references"><B>References</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/BIB_NeedNLM.sql.rpt">References Which Need NLM Updates</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/BIB_InPress.sql.rpt">References In Press/Submissions</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/BIB_NoAbstract.sql.rpt">References w/ Medline UI and No Abstract</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/BIB_Submission.sql.rpt">Potential Submission Matches</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/BIB_GBSubNoMarker.sql.rpt">GenBank Submission References with no Marker associations</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/BIB_GBSubMatch.sql.rpt">Markers w/ GenBank Submission References which match non-GenBank Submission References</A>
</OL>

<A NAME="markers"><B>Markers</B> </A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_OffsetsMLDP.sql.rpt">Markers with No Offsets Modifed in Mapping w/in the last 3 days</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Offsets.sql.rpt">Markers w/ MGD Offset but w/out CC Offset or Mapping Information</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_OffsetsUN.sql.rpt">Markers with Offsets but no Chromosome Assignment</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Alias.sql.rpt">Aliases</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_NoReferences.sql.rpt">Markers w/out ANY References</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_References.sql.rpt">Markers w/out History References or History</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_WithdrawnAdditional.sql.rpt">Splits and Withdrawn Markers w/ Additional Info</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Matrix.sql.rpt">UN/Syntenic Markers In MGD found in Mapping Panel Crosses</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_AccDup.sql.rpt">Duplicate Marker Accession Numbers (Sequence, SWISS-PROT)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_UniGeneClusterDiscrep.sql.rpt">Clusters with Markers on different chromosomes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_LowerSeq.rpt">Lowercase Gene Symbols which have GenBank IDS</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_AssigningRef.rpt">MGI Markers with Assigning Reference</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_MGC.sql.rpt">MGC IDs with Nomen Records Not Deleted or Broadcast</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_NoSequence.rpt">Genes with No Sequence Data</A>
</OL>

<A NAME="alleles"><B>Alleles</B> </A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Alleles.sql.rpt">Allele Anomalies</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/ALL_NomenText.sql.rpt">Symbols which have undergone withdrawals and appear in Allele Molecular Notes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/ALL_ESCellStrain.sql.rpt">ES Cell Line Strains Modified Last Month - Sorted by Modification Date</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/ALL_ReservedNot23000.sql.rpt">Reserved Alleles w/ Primary Reference != J:23000</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_WithAllele_MLC.sql.rpt">Allele Withdrawals in Last 2 Months where either Old or New Gene has MLC record</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/ALL_Progress.sql.current.rpt">Allele Progress Report</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/archive/alleles">Allele Progress Archive</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/ALL_NoMutantSentence.rpt">Genes w/Phenotypic Alleles and No Mutant Sentence</A>
</OL>

<A NAME="mlc"><B>MLC</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_MLC_Withdrawn.sql.rpt">Withdrawn Markers w/ MLC Text</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLC_Alleles.sql.rpt">MLC Symbols which contain Alleles</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLC_References.sql.rpt">References which only appear in MLC</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLC_CheckImageIndex.rpt">MLC Image Index File Entries Which Contain Withdrawn Symbols</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLC_TransferStats.rpt">MLC Transfer Statistics Report</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLCNomenclature-current.html">Updates to MLC Mouse Nomenclature</A>
</OL>

<A NAME="probes"><B>Probes</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_Reference.sql.rpt">Probes/Primers w/out References (excluding I.M.A.G.E. clones)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_Marker.sql.rpt">Probes Used More Than Once for Marker/Duplicate Markers/No Markers (excluding I.M.A.G.E. clones)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_NoGenBank.sql.rpt">No GenBank IDs - Sequence in Paper</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_NoMarkerSeq.sql.rpt">Molecular Segments with Encoding Markers and Sequence IDs and no corresponding Marker-Sequence ID association</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_MarkerSeq.sql.rpt">I.M.A.G.E. Clone/Marker Sequence Anomalies</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MGI_ProblemSequence.rpt">Molecular Segment with the Problem Sequence Note associated with more than one GenBank Seqid or associated with no GenBank Seqid</A>
<LI><A HREF="ftp://ftp.informatics.jax.org/pub/informatics/reports/MGI_ProblemSequence.rpt">Public Problem Sequence Report</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_RIKENMarkers.rpt">RIKEN Clones and MGI Markers</A>
</OL>

<A NAME="mapping"><B>Mapping</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_Chromosome.sql.rpt">Invalid Chromosomes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_2Point.sql.rpt">Invalid Recombinants vs. Parentals/Recombinants vs. Total</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_Cross.sql.rpt">Invalid Cross Data</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_Marker.sql.rpt">Missing Experimental Markers</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_DupMarker.sql.rpt">Duplicate Markers Appearing in Experiments</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_ExptMismatch.sql.rpt">Experiments which are of one Experiment Type but which have corresponding data records in non-related Experiment tables</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_NomenText.sql.rpt">Symbols which have undergone withdrawals and appear in Text Experiments</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_MappedNoRef.sql.rpt">Mapped Markers w/out a Mapping Reference</A>
</OL>

<A NAME="homology"><B>Homology</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_References2.sql.rpt">References selected for Homology only w/out Homology Data</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_References5.sql.rpt">References selected for Homology w/ no Homology Data with at least one reference-to-gene Association</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_SpeciesCount.sql.rpt">Counts of Mouse Homologies by Species</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_Duplicates.sql.rpt">Duplicate Homologies (by Class, Reference, Species)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_ConservedUN.sql.rpt">Homologies w/ Conserved Map Location and Chromosome = UN</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_NoMouseSeq.sql.rpt">Mouse Markers in Homology w/ Nucleotide or Amino Acid Sequence Similarity Assay and Marker has no Seq ID</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_RIKEN.sql.rpt">Human/Mouse Homologies involving RIKEN Symbols</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_MouseSeq1.sql.rpt">Mouse Genes with Sequence ID but no Human Homology</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_MouseSeq2.sql.rpt">Mouse Genes with no Sequence ID but with Human Homology</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_NoMouse.sql.rpt">Homology Records w/out Mouse Genes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_LocusLink1.sql.rpt">LocusLink Mouse/Human Orthologous Pairs where we have neither the mouse nor the human member of the pair. (HUMAN AND MOUSE SYMBOLS ARE THE SAME)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_LocusLink2A.sql.rpt">LocusLink Mouse/Human Orthologous Pairs where we have neither the mouse nor the human member of the pair. (HUMAN AND MOUSE SYMBOLS ARE NOT THE SAME, HUMAN SYMBOLS ARE OFFICIAL)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_LocusLink2B.sql.rpt">LocusLink Mouse/Human Orthologous Pairs where we have neither the mouse nor the human member of the pair. (HUMAN AND MOUSE SYMBOLS ARE NOT THE SAME, HUMAN SYMBOLS ARE INTERIM)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_LocusLink3.sql.rpt">LocusLink Mouse/Human Orthologous Pairs where our Orthology contains Human but no Mouse member.</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_LocusLink4.sql.rpt">LocusLink Mouse/Human Orthologous Pairs where our Orthology contains Mouse but no Human member.</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_GDBLocusLink.sql.rpt">Human Markers with > 1 GDB record ID in LocusLink</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_DupGDB.sql.rpt">Human Markers with > 1 GDB record ID</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_Singletons.sql.rpt">Mouse/Human Singletons</A>
</OL>

<A NAME="gxd"><B>GXD</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_ProbeMarker.sql.rpt">GXD Assay Probe/Marker Pairs No Longer Found in Master Probe Table</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_AntibodyMarker.sql.rpt">GXD Assay Antibody/Marker Pairs No Longer Found in Master Antibody Table</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_ProbeReference.sql.rpt">GXD Assay Probes with no corresponding entry in the Probe Reference Table</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_AltTranscripts.sql.rpt">Alterntative Transcripts that have "encodes" relationship for more than one gene</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_NotInCache.sql.rpt">GXD Assays missing from GXD Expression Cache Table</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_Stats.sql.rpt">GXD Statistics</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_NotGene.sql.rpt">GXD Assays and Index with Markers which are not Genes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_SpecEmbryonic.sql.rpt">Adult Specimens with Embryonic Structures</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_SpecNoAge.sql.rpt">Specimens with NULL Age Value (when there should be one)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_SpecMissingAllele.sql.rpt">Specimens with Missing Allele 2 Genotypes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_SpecTheiler.sql.rpt">Specimens annotated to structures of > 1 Theiler Stage</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_MechDev.sql.rpt">Mech Dev papers to full code</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GXD_GeneExprPatterns.sql.rpt">Gene Expr Patterns papers to full code</A>
</OL>

<A NAME="go"><B>GO</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_RatNoGO.sql.rpt">Mouse Genes that have Rat Homologs but no GO associations</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_NoGO_A.rpt">Genes (neither 'RIKEN' or 'expressed' or 'EST') with no GO Associations and reference count excludes data loads and GenBank submissions</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_NoGO_B.rpt">Genes (neither 'RIKEN' or 'expressed' or 'EST') with no GO Associations</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_NoGO_C.rpt">Genes (neither 'RIKEN' or 'expressed' or 'EST') with no GO Associations and reference count includes only data loads and GenBank submissions</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_NoGO_D.rpt">Genes (neither 'RIKEN' or 'expressed' or 'EST') with no GO Associations with references that are selected for GO but have not been used</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_GOIEA_A.rpt">Genes (neither 'RIKEN' or 'expressed' or 'EST') with GO Associations of evidence IEA only and reference count excludes data loads and GenBank submissions</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_GOIEA_B.rpt">Genes (neither 'RIKEN' or 'expressed' or 'EST') with GO Associations of evidence IEA only</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_GOIEA_C.rpt">Genes (neither 'RIKEN' or 'expressed' or 'EST') with GO Associations of evidence IEA only and reference count includes only data loads and GenBank submissions</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_GOGold_A.rpt">All Genes with Annotations to IDA, IGI, IMP, IPI, TAS</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_GOGold_B.rpt">All Genes with Annotations to IDA, IGI, IMP, IPI</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_GOGold_C.rpt">All Genes with multiple Annotations to IDA, IGI, IMP, IPI, TAS</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_GOGold_D.rpt">All Genes with multiple Annotations to IDA, IGI, IMP, IPI</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_GOAllele.rpt">Genes with Annotation to IMP, Inferred From = NULL and Reference Associations to Alleles</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GO_EvidenceCodeStats.sql.rpt">Total # of Annotations by GO Evidence Code</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/GO_stats.rpt">GO Statistics</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/archive/go">Archive of GO Statistics</A>
</OL>

<A NAME="fantom"><B>Fantom2</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/FAN2_IncompleteCluster.rpt">Incomplete Cluster</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/FAN2_InvalidFinalMGIID.rpt">Invalid Final MGI ID</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/FAN2_MultRepClone.rpt">Multiple Representative Clones</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/FAN2_NoRepClone.rpt">No Representative Clones</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/FAN2_NonRikenRepClone.rpt">non-RIKEN Representative Clones</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/FAN2_DiscrRepClone.rpt">Representative Clone Discrepancies</A>
</OL>

<A NAME="nomenclature"><B>Nomenclature</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLP_Nomenclature.sql.rpt">Strains Affected by recent Nomenclature changes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/NOMEN_Broadcast.sql.rpt">Nomenclature Symbols Broadcast Within Last 3 Days</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/NOMEN_Pending.sql.rpt">Nomenclature In Progress Symbols</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/NOMEN_Duplicate.sql.rpt">Nomenclature Duplicate Symbols</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/NOMEN_Reserved.rpt">Nomenclature Reserved</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Duplicates.sql.rpt">MGD Duplicate Markers</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/HMD_SymbolDiffs.sql.rpt">MGD Mouse Symbols which differ from Orthologous Human Symbols</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Withdrawn.sql.rpt">Withdrawn Markers</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_DashPending.sql.rpt">-pending Markers</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_NomenTransgene.rpt">Transgenes with nomenclature changes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/ALL_NomenTransgene.rpt">Transgene Alleles  with nomenclature changes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_NomenRat.sql.rpt">Mouse Symbols with nomenclature changes which contain Rat homologs with unofficial nomenclature</A>
</OL>

<A NAME="strains"><B>Strains</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_StrainStd.sql.rpt">Strains (Standard only)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_StrainNonStd.sql.rpt">Strains (Non Standard excluding F1 and F2)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_StrainF1F2.sql.rpt">Strains Containing F1 or F2</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_StrainDup.sql.rpt">Duplicate Strains</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_StrainByDate.sql.rpt">Strains (Non Standard - Sorted by Creation Date)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_Strain_NameChanges.rpt">Strain Names Which Have Been Changed</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_Strain.rpt">Tab-delimited File of All Strains (MGI Acc ID, Strain, Markers, Synonyms, External Acc IDs)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_Strain2.rpt">Tab-delimited File of All Strains (Strain/Standard)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_StrainTrans.sql.rpt">Strains - Transgenic or Targeted Mutation</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_StrainNonStdNoData.sql.rpt">Non-Standard Strains with no JR# and no data attached</A>
</OL>

<A NAME="dataloads"><B>Data Loads</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/dataload/locuslinkload/LLindex.html">LocusLink Load: Data and QC Reports</A>
<LI><A HREF="/usrlocalmgi/dataload/ratmapload/RATMAPindex.html">RATMAP Load: Data and QC Reports</A>
<LI><A HREF="/usrlocalmgi/dataload/swissload">SwissProt Load Data and Logs</A>
<LI>UniGene</LI>
	<OL>
	<LI><A HREF="/usrlocalmgi/qcreports_db/output/multipleClusters.rpt">MGI Genes associated with multiple UniGene Clusters</A>
	<LI><A HREF="/usrlocalmgi/qcreports_db/output/multipleGenes.rpt">UniGene Clusters associated with multiple MGI Genes</A>
	<LI><A HREF="/usrlocalmgi/qcreports_db/output/genesMissingUG.rpt">MGI mouse Genes with Seq. ID but not in a UniGene Cluster</A>
	</OL>
<LI><A HREF="/usrlocalmgi/dataload/crossload/crossloadindex.html">CrossLoad: Data and QC Reports</A>
<LI><A HREF="/usrlocalmgi/dataload/vocload/GOindex.html">GO Vocabulary/DAG Load</A>
<LI><A HREF="/usrlocalmgi/dataload/expseqload/index.html">RH-Mapped Expressed Sequence Load (Lucy Load)</A>
</OL>


<A NAME="lists"><B>Lists</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Species.sql.rpt">Marker Species</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MRK_Classes.sql.rpt">Marker Classes</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_Assays.sql.rpt">Mapping Assays</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/MLD_Crosses.sql.rpt">Mapping Panels Displayed in Lookup List</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_Tissue.sql.rpt">Tissues (All)</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_CellLine.sql.rpt">Cell Lines</A>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/PRB_TissueCell.sql.rpt">Unique Tissue/Cell Lines Pairs</A>
</OL>

<A NAME="other"><B>Other</B></A>
<OL>
<LI><A HREF="/usrlocalmgi/qcreports_db/output/ACC_Accession.sql.rpt">Records missing MGI Accession Numbers</A>
</OL>

<HR>

</BODY>
</HTML>
